contracts/FungibleTokenMetadataViews.cdc

```diff
@@ -0,0 +1,185 @@
+            self.logos = logos
+            self.socials = socials
+        }
+    }
+
+    /// Helper to get FTDisplay in a way that will return a typed optional.
+    /// 
+    /// @param viewResolver: A reference to the resolver resource
+    /// @return An optional FTDisplay struct
+    ///
+    pub fun getFTDisplay(_ viewResolver: &{MetadataViews.Resolver}): FTDisplay? {
+        if let maybeDisplayView = viewResolver.resolveView(Type<FTDisplay>()) {
+            if let displayView = maybeDisplayView as? FTDisplay {
+                return displayView
+            }
+        }
+        return nil
+    }
+
+    /// View to expose the information needed store and interact with a FT vault.
+    /// This can be used by applications to setup a FT vault with proper 
+    /// storage and public capabilities.
+    ///
+    pub struct FTVaultData {
+        /// Path in storage where this FT vault is recommended to be stored.
+        pub let storagePath: StoragePath
+
+        /// Public path which must be linked to expose the public receiver capability.
+        pub let receiverPath: PublicPath
+
+        /// Public path which must be linked to expose the balance and resolver public capabilities.
+        pub let metadataPath: PublicPath
+
+        /// Private path which should be linked to expose the provider capability to withdraw funds 
+        /// from the vault.
+        pub let providerPath: PrivatePath
+
+        /// Type that should be linked at the `receiverPath`. This is a restricted type requiring 
+        /// the `FungibleToken.Receiver` interface.
+        pub let receiverLinkedType: Type
+
+        /// Type that should be linked at the `receiverPath`. This is a restricted type requiring 
+        /// the `FungibleToken.Balance` and `MetadataViews.Resolver` interfaces.
+        pub let metadataLinkedType: Type
+
+        /// Type that should be linked at the aforementioned private path. This 
+        /// is normally a restricted type with at a minimum the `FungibleToken.Provider` interface.
+        pub let providerLinkedType: Type
+
+        /// Function that allows creation of an empty FT vault that is intended
+        /// to store the funds.
+        pub let createEmptyVault: ((): @FungibleToken.Vault)
+
+        init(
+            storagePath: StoragePath,
+            receiverPath: PublicPath,
+            metadataPath: PublicPath,
+            providerPath: PrivatePath,
+            receiverLinkedType: Type,
+            metadataLinkedType: Type,
+            providerLinkedType: Type,
+            createEmptyVaultFunction: ((): @FungibleToken.Vault)
+        ) {
+            pre {
+                receiverLinkedType.isSubtype(of: Type<&{FungibleToken.Receiver}>()): "Receiver public type must include FungibleToken.Receiver."
+                metadataLinkedType.isSubtype(of: Type<&{FungibleToken.Balance, MetadataViews.Resolver}>()): "Metadata public type must include FungibleToken.Balance and MetadataViews.Resolver interfaces."
```

Comment:
should this also ensure that it is a subtype of receiver? personally I never understood why FT capabilities do not link both receiver and balance.

Reply:
That's a good question. I guess that if we are making sure that a `{FungibleToken.Receiver}` is provided on the `receiverLinkedType` we don't need to check for it on the `metadataLinkedType`.

Also you make me thing if we really want to be checking that the `providerLinkedType` is also a `{MetadataViews.Resolver}`